mongo/session.go

```diff
@@ -281,6 +284,9 @@ func (s *sessionImpl) StartTransaction(opts ...*options.TransactionOptions) erro
 
 // AbortTransaction implements the Session interface.
 func (s *sessionImpl) AbortTransaction(ctx context.Context) error {
```

Comment:
Sessions are documented as not safe for use by multiple goroutines:

This is documented in the [Go driver](https://github.com/mongodb/mongo-go-driver/blob/c29474b05cce7c75138343704848d35943d71038/mongo/session.go#L34):

> Implementations of SessionContext are not safe for concurrent use by multiple goroutines.

And [the specifications](https://github.com/mongodb/specifications/blob/745e486dd03f0d724c68593bf9ddb017d2d58fa6/source/sessions/driver-sessions.rst):

> ClientSession instances are not thread safe or fork safe. They can only be used by one thread or process at a time.

> Drivers MUST document the thread-safety and fork-safety limitations of sessions. Drivers MUST NOT attempt to detect simultaneous use by multiple threads or processes (see Q&A for the rationale).

Does this suggest the cause of this crash is incorrect use of the session across multiple goroutines? If so, that may suggest a bug in the application.

Reply:
> Does this suggest the cause of this crash is incorrect use of the session across multiple goroutines? 

This would be my guess. The only way I could re-create this bug is asynchronously aborting transactions, I don't think there is any other way to get the "nil pointer dereference" error at "x/mongo/driver/session/client_session.go:317". [Here](https://github.com/10gen/baas/blob/master/api/common/graphql.go#L356) is where BAAS is aborting this transaction, which is part of a promise pattern defined [here](https://github.com/10gen/baas/blob/master/api/common/graphql.go#L395).

The reason I am proposing this change is that the specifications also say

> If this session is already in the "transaction aborted" state, then drivers MUST raise an error containing the message "Cannot call abortTransaction twice".

It appears that when you call "abortTransaction" asynchronously, the state may change between instructions in the "[Client.ClearedPinnedResources()](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/session/client_session.go#L315)" method. 

Even though we document that implementations of SessionContext are not safe, should we not guard state changes to avoid "nil pointer dereference" runtime errors?

Reply:
The panic may suggest a SessionContext is used concurrently by multiple goroutines. Consider asking the BAAS team to run a test that exercises the code path triggering the abort using the race detector.

Another option is to make a best-effort attempt to detect incorrect concurrent use of the SessionContext and return a more helpful error. This is [prohibited by the specification](https://github.com/mongodb/specifications/blob/745e486dd03f0d724c68593bf9ddb017d2d58fa6/source/sessions/driver-sessions.rst#why-do-we-say-drivers-must-not-attempt-to-detect-unsafe-multi-threaded-or-multi-process-use-of-clientsession). But IMO a simple best-effort check may help to detect concurrent use and provide an easier to diagnose error. Here is [one proposal](https://github.com/mongodb/mongo-go-driver/compare/master...kevinAlbs:mongo-go-driver:session_safeguard?expand=1).